bind.go

```diff
@@ -189,6 +213,86 @@ func (ml *maybeLineage) checkLensesOrder() error {
 	return nil
 }
 
+func (ml *maybeLineage) checkGoLensCompleteness() error {
+	all := make(map[lensID]bool)
+	for _, lens := range ml.implens {
+		all[lid(lens.To, lens.From)] = true
+	}
+
+	var missing []lensID
+
+	var prior SyntacticVersion
+	for _, sch := range ml.schlist[1:] {
+		// there must always at least be a reverse lens
+		v := sch.Version()
+		revid := lid(prior, v)
+
+		if !all[revid] {
+			missing = append(missing, revid)
+		} else {
+			delete(all, revid)
+		}
+
+		if v[0] != prior[0] {
+			// if we crossed a major version, there must also be a forward lens
+			fwdid := lid(v, prior)
+			if !all[fwdid] {
+				missing = append(missing, fwdid)
+			} else {
+				delete(all, fwdid)
+			}
+		}
+		prior = v
+	}
+
+	// TODO is it worth making each sub-item into its own error type?
+	if len(missing) > 0 {
+		b := new(bytes.Buffer)
+
+		fmt.Fprintf(b, "Go migrations not provided for the following version pairs:\n")
+		for _, mlid := range missing {
+			fmt.Fprint(b, "\t", mlid, "\n")
+		}
+		return errors.Mark(errors.New(b.String()), terrors.ErrMissingLenses)
+	}
+
+	if len(all) > 0 {
+		b := new(bytes.Buffer)
+
+		fmt.Fprintf(b, "Go migrations erroneously provided for the following version pairs:\n")
+		// walk the slice so output is reliably ordered
+		for _, lens := range ml.implens {
+			// if it's not in the list it's because it was expected & already processed
+			elid := lid(lens.To, lens.From)
+			if _, has := all[elid]; !has {
+				continue
+			}
+			if !synvExists(ml.allv, lens.To) {
+				fmt.Fprintf(b, "\t%s (schema version %s does not exist", elid, lens.To)
+			} else if !synvExists(ml.allv, lens.From) {
+				fmt.Fprintf(b, "\t%s (schema version %s does not exist", elid, lens.From)
+			} else if elid[0] == elid[1] {
+				fmt.Fprintf(b, "\t%s (self-migrations not allowed)", elid)
+			} else if elid[0].Less(elid[1]) {
```

Comment:
Nit; I usually find a bit difficult to reason about _to->from->to_ with slice indexes (_[0], [1]..._), and even more considering I'm a LTR reader and in Thema we usually define lenses in the opposite order.

So, for the sake of making this cases a bit easier to follow and reason about, I think we could get benefit of the type introduce in this pull request `type lensID ...` to add some semantic methods for these checks (`elid[0] == elid[1]`, elid[0].Less(elid[1])`, etc).

Wdyt?

Reply:
yeah, i was having a similar thought while writing this and getting lost in the logic. A couple things occurred to me:

* Make it a struct with explicit fields rather than positional in an array
* The typical way of printing this is LTR, being `<From> -> <To>` - perhaps that's also how the fields should be ordered

Reply:
just came to this now after writing the same comments above